### ISSUE TITLE
GetField and monitor give different structures

### ISSUE THREAD

**Issue by mdavidsaver** [epics-base/pvaSrv#38]
I find that the structure description returned by GetField and Monitor differ slightly.  The valueAlarm sub-struct is missing in one.  So ```pvinfo <pvname>``` doesn't show the same structure as ```pvget -r "" -m <pvname>```.  I had thought that they must be the same.

Is this a bug in code or my expectations?

**Comment by dhickin** [comment]
Bug in code. pvinfo and pvget -r "" should return the same structure. To be clear the valueAlarm field is missing in pvinfo and present in pvget.

Note
1. The normative type NTScalar doesn't have the valueAlarm as either a mandatory or optional field (although that doesn't prevent a conformant NTScalar from having it as an extra field).
2. The corresponding behaviour in caProvider is to provide the valueAlarm field in both cases.

Probably easiest fix is to add valueAlarm to GetField

**Comment by dhickin** [comment]
The issue is somewhat more complicated.

I can see 3 basic cases for scalars and 3 for arrays.

I would expect a GetField to return the same fields as a get or monitor of all fields and I would hope ideally to get the same for dbPV and caProvider.

Comparing the behaviour of pvinfo and pvget -r "field()" over pva and ca for testTop PVs and listing the presence of display (d), control (c) and valueAlarm (vA) fields we get

For scalars

    pv | rtype | info | get | info (ca) | get (ca) | case
    ---| ------| ---- | --- | --------- | -------- | -----
    double01 | ai       |d,c| d,c,vA | d,c,vA | d,c,vA | 1
    int01   | longout   |d,c| d,c,vA | d,c,vA | d,c,vA | 1
    string01| stringout |d,c|        |        |        | 2
    *.RTYP  |           |d,c|        |        |        | 2
    byte01  | byter     |d,c|        | d,c,vA | d,c,vA | 3
    short01 | shortr    |d,c|        | d,c,vA | d,c,vA | 3

and for arrays

    pv | rtype | info | get | info (ca) | get (ca) | case
    ---| ------| ---- | --- | --------- | --------  | -----
    doubleArray01 | waveform |d | d,c,vA | d,c,vA | d,c,vA | 4
    intArray01    | waveform |d | d,c,vA | d,c,vA | d,c,vA | 4
    shortArray01  | waveform |d |        |        |        | 5
    stringArray01 | waveform |d |        |        |        | 5
    byteArray01   | waveform |d |        | d,c,vA | d,c,vA | 6
    shortArray01  | waveform |d |        | d,c,vA | d,c,vA | 6

So correct behaviour ought to include info column matching get column (and ideally -p ca columns) . In other words the behaviour does not appear to be correct for any of these PVs.

So I can see 3 cases for scalars and 3 for arrays

1) for ao and longout records double01 and int01 the GetField response is missing valueAlarm that is present in the channel get/monitor response and in getField and channel get/monitor for caProvider. Easy fix is to add this by adding valueAlarm field. Note also that the valueAlarm fields corresponding to HHSV, HSV, LSV, LLSV, HYST are not filled in.

2) For strings stringout and *.rtyp the returned structure for GetField has display and control fields, but the channel get/monitor fields do not. Similarly caProvider does not return these fields. Also the fix for a above causes all string gets to fail (as an exception is thrown). Easy fix is to remove control and display (and valueAlarm) fields for strings.  This makes GetField match channel get and caProvider operations and string gets succeed.

3) For byte01 and short01 of type byter, shortr the GetField type has display and control fields, but get does not have these.  caProvider operations have control, display and valueAlarm. I'll have to look further into a fix for this case. The above suggested fixes will at least bring the GetField response in line with caProvider structure.

4) Array version of 1. GetField has display field only. This seems reasonable as control is not an optional field of NTScalarArray (so if added is an extra field) and alarmLimit isn't either, so we might not want to add valueAlarm. Channel/Get returns display, control and valueAlarm as do caProvider operations. The returned type ought to be the same whichever way we go.

5) Array version of 2. GetField has display field. Channel Get/monitor and caProvider operations do not. Fix along the lines of 2 will fix this.

6) Array version of 3. Again I'll have to think about this.

I suggest adding the valueAlarm in the GetField response for scalars which are not strings and removing the control and display fields for strings and removing display for string arrays.

That gives correct behaviour for 1,2 and 5. GetField structure will match channel get structure and caProvider behaviour.

For case 4 I can bring behaviour in line with get and caProvider, at the expense of adding extra fields to the normative type. This is easy. Alternative is to change the get/monitor behaviour (and caProvider too).
@ralphlange do you have a view on this?

**Comment by dhickin** [comment]
I now see why cases 3 and 6 happen. In 3.15 version:
```C++
        if(fieldList.find(displayString)!=string::npos) {
            if(dbAddr.field_type==DBF_LONG||dbAddr.field_type==DBF_DOUBLE) {
                propertyMask |= displayBit;
            }
        }
```
and similarly for 3.14.

Is there a reason why display, control and alarm fields are removed for CHARs,SHORTs, FLOATs,and their unsigned versions?

If we replace these with
```C++
        switch(dbChannelFinalDBFType(dbChan))
        {
        case DBF_CHAR:
        case DBF_UCHAR:
        case DBF_SHORT:
        case DBF_USHORT:
        case DBF_LONG:
        case DBF_ULONG:
        case DBF_FLOAT:
        case DBF_DOUBLE:
            if(fieldList.find(displayString)!=string::npos) {
                propertyMask |= displayBit;
            }
            if(fieldList.find(controlString)!=string::npos) {
                propertyMask |= controlBit;
            }
            if(fieldList.find(valueAlarmString)!=string::npos) {
                propertyMask |= valueAlarmBit;
            }
            break;
        default:
            break;
        }
``` 
this will fix case 3. If we also remove control and valueAlarm for scalar array PVs:
```C++
        switch(dbChannelFinalDBFType(dbChan))
        {
        case DBF_CHAR:
        case DBF_UCHAR:
        case DBF_SHORT:
        case DBF_USHORT:
        case DBF_LONG:
        case DBF_ULONG:
        case DBF_FLOAT:
        case DBF_DOUBLE:
            if(fieldList.find(displayString)!=string::npos) {
                propertyMask |= displayBit;
            }
            if(type==scalar && fieldList.find(controlString)!=string::npos) {
                propertyMask |= controlBit;
            }
            if(type==scalar && fieldList.find(valueAlarmString)!=string::npos) {
                propertyMask |= valueAlarmBit;
            }
            break;
        default:
            break;
        }
``` 
then this will fix all cases (there will be some difference in behaviour with caProvider - caProvider scalar arrays will have control and valueAlarm fields- but we can change the latter).

I can add these changes to my pull request. Can anyone suggest a reason not to do this?

**Comment by ralphlange** [comment]
Please go ahead.
What you propose sounds absolutely reasonable.

After taking over the code, I was looking at some of this, scratching my head, and realizing this was probably not the way it should be.
However, in the end I didn't dare to change anything, because with my limited insight I was too scared to break things, and there were no tests that I could rely upon for finding unwanted side effects of changes.

**Comment by dhickin** [comment]
Thank Ralph.

OK, I've added the changes to my pull request. See what you think.

**Comment by dhickin** [comment]
Fixed by #40.
